File: iosApp/Shared/UI/DetailScreen.swift
//
//  DetailScreen.swift
//  iosApp
//
//  Created by Robert Zetzsche on 28.10.23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared
import SwiftUI
import KMMViewModelSwiftUI

@available(iOS 15.0, *)
struct DetailScreen: View {
    let bluetoothDevice: BluetoothDevice
    @StateViewModel var viewModel: DetailViewModel
    
    init(bluetoothDevice: BluetoothDevice) {
        self.bluetoothDevice = bluetoothDevice
        let detailViewModel: DetailViewModel = getKoinInstance(parameters: [bluetoothDevice])
        self._viewModel = StateViewModel(wrappedValue: detailViewModel)
    }
    
    var body: some View {
        
        VStack{
            Text(bluetoothDevice.name ?? "Air Tag")
            Text(bluetoothDevice.identifier)
            Text(bluetoothDevice.rssi!.description)
            Spacer()
            
            switch onEnum(of: viewModel.state) {
            case .loading(_):
                ProgressView()
            case .loaded(let data):
                switch  onEnum(of: data) {
                case.airTag(let loaded):
                    Text("Battery Level: \(loaded.batteryLevel)")
                    if(viewModel.state is DetailViewModel.StateLoadedAirTagLoaded){
                        Button(action: {viewModel.playAirtagSound()}, label: {
                            Text("Play Sound")
                        })
                    }else{
                        ProgressView()
                    }
                case .movisensSensorLoaded(let loaded):
                    let values =   loaded.movisensSensorValues
                    Text("\(values.manufacturer) \(values.modelNumber)")
                    Text("Firmware Version: \(values.firmware)")
                    Text("Serial Number: \(values.serialNumber)")
                    Text("Battery Level: \(values.batteryLevel) %")
                }
                
            case .error(let error):
                Text(error.error)
            }
            
        }.padding(EdgeInsets(top: 32, leading: 16, bottom: 16, trailing: 16))
    }
}

File: sharedCompose/src/commonMain/kotlin/com/movisens/bluetooth/kmm/view/DetailScreen.kt
package com.movisens.bluetooth.kmm.view

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material.Button
import androidx.compose.material.CircularProgressIndicator
import androidx.compose.material.Icon
import androidx.compose.material.IconButton
import androidx.compose.material.Scaffold
import androidx.compose.material.Text
import androidx.compose.material.TopAppBar
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import cafe.adriel.voyager.core.screen.Screen
import cafe.adriel.voyager.navigator.LocalNavigator
import cafe.adriel.voyager.navigator.currentOrThrow
import com.movisens.bluetooth.kmm.repository.bluetooth.BluetoothDevice
import com.movisens.bluetooth.kmm.viewmodel.DetailViewModel
import com.movisens.bluetooth.kmm.viewmodel.DetailViewModel.State.Error
import com.movisens.bluetooth.kmm.viewmodel.DetailViewModel.State.Loaded.AirTag
import com.movisens.bluetooth.kmm.viewmodel.DetailViewModel.State.Loading
import org.koin.core.component.KoinComponent
import org.koin.core.parameter.parametersOf

class DetailScreen(private val bluetoothDevice: BluetoothDevice) : Screen, KoinComponent {
    private val detailViewModel =
        getKoin().get<DetailViewModel>(parameters = { parametersOf(bluetoothDevice) })

    @Composable
    override fun Content() {
        val detailState by detailViewModel.state.collectAsState()
        val navigator = LocalNavigator.currentOrThrow
        Scaffold(topBar = {
            TopAppBar(title = { Text("App") }, navigationIcon = {
                IconButton(onClick = { navigator.pop() }) {
                    Icon(Icons.Default.ArrowBack, null)
                }
            })
        }) {

            Column(
                modifier = Modifier.padding(16.dp).fillMaxWidth(),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Text(bluetoothDevice.name ?: "Air Tag")
                Text(bluetoothDevice.identifier)
                Text(bluetoothDevice.rssi.toString())
                Box(modifier = Modifier.height(100.dp))

                detailState.let { detailState ->
                    when (detailState) {
                        is DetailViewModel.State.Loaded.MovisensSensorLoaded -> Column {
                            val movisensSensorValues = detailState.movisensSensorValues
                            Text("${movisensSensorValues.manufacturer} ${movisensSensorValues.modelNumber}")
                            Text("Firmware Version: ${movisensSensorValues.firmware}")
                            Text("Serial Number: ${movisensSensorValues.serialNumber}")
                            Text("Battery Level: ${movisensSensorValues.batteryLevel} %")
                        }

                        is Error -> Text(detailState.error)
                        is AirTag -> {
                            Text("Battery Level: ${detailState.batteryLevel}")
                            Box(modifier = Modifier.height(100.dp))
                            if (detailState is AirTag.Loaded)
                                Button(onClick = { detailViewModel.playAirtagSound() }) {
                                    Text("Play Sound")
                                }
                            else {
                                CircularProgressIndicator()
                            }
                        }

                        is Loading -> CircularProgressIndicator()
                    }
                }
            }
        }
    }
}